### ISSUE TITLE
torch.compile produces inconsistent outputs for Multi-Query Attention model with manually implemented attention

### ISSUE THREAD

**Issue by Blooming-Tree** [pytorch/pytorch#188432]
### 🐛 Describe the bug

When applying torch.compile to a Multi-Query Attention (MQA) model with manually implemented attention computation (q @ k.transpose(-2, -1) + F.softmax), the compiled model produces outputs that significantly deviate from eager execution, exceeding acceptable error thresholds.
code:
```
import os
import copy
import torch
import torch.nn as nn
import torch.nn.functional as F

class MultiQueryAttention(nn.Module):
    def __init__(self, embed_dim, num_heads, dropout=0.1):
        super().__init__()
        self.embed_dim = embed_dim
        self.num_heads = num_heads
        self.head_dim = embed_dim // num_heads
        assert self.head_dim * num_heads == embed_dim
        self.q_proj = nn.Linear(embed_dim, embed_dim)
        self.k_proj = nn.Linear(embed_dim, self.head_dim)
        self.v_proj = nn.Linear(embed_dim, self.head_dim)
        self.out_proj = nn.Linear(embed_dim, embed_dim)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x, mask=None):
        batch_size, seq_len, _ = x.shape
        q = self.q_proj(x)
        k = self.k_proj(x)
        v = self.v_proj(x)
        q = q.view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
        k = k.unsqueeze(1).expand(-1, self.num_heads, -1, -1)
        v = v.unsqueeze(1).expand(-1, self.num_heads, -1, -1)
        attn_scores = self.head_dim ** 0.5 / (q @ k.transpose(-2, -1))
        if mask is not None:
            attn_scores = attn_scores.masked_fill(mask == 0, float('-inf'))
        attn_weights = F.softmax(attn_scores, dim=-1)
        attn_weights = self.dropout(attn_weights)
        output = attn_weights @ v
        output = output.transpose(1, 2).contiguous().view(batch_size, seq_len, self.embed_dim)
        output = self.out_proj(output)
        return output

class MQA_DNN(nn.Module):
    def __init__(self, embed_dim, num_heads, hidden_dim, num_layers, dropout=0.1):
        super().__init__()
        self.layers = nn.ModuleList([
            nn.ModuleDict({
                'mqa': MultiQueryAttention(embed_dim, num_heads, dropout),
                'ffn': nn.Sequential(
                    nn.Linear(embed_dim, hidden_dim),
                    nn.ReLU(),
                    nn.Dropout(dropout),
                    nn.Linear(hidden_dim, embed_dim),
                    nn.Dropout(dropout)
                ),
                'norm1': nn.LayerNorm(embed_dim),
                'norm2': nn.LayerNorm(embed_dim)
            }) for _ in range(num_layers)
        ])

    def forward(self, x, mask=None):
        for layer in self.layers:
            attn_output = layer['mqa'](layer['norm1'](x), mask)
            x = x + attn_output
            ffn_output = layer['ffn'](layer['norm2'](x))
            x = x + ffn_output
        return x

def get_default_model():
    return MQA_DNN(embed_dim=512, num_heads=8, hidden_dim=2048, num_layers=6, dropout=0.1)

def get_sample_inputs():
    return (torch.randn(2, 20, 512), None)

def main():
    torch.use_deterministic_algorithms(True)
    torch.set_float32_matmul_precision('highest')
    os.environ['CUBLAS_WORKSPACE_CONFIG'] = ':4096:8'
    torch._inductor.config.fallback_random = True

    torch.manual_seed(42)
    model = get_default_model()
    model = model.double()
    model.eval()

    inputs = get_sample_inputs()
    inputs = tuple(t.double() if isinstance(t, torch.Tensor) and t.is_floating_point() else t for t in inputs)

    with torch.no_grad():
        eager_output = model(*inputs)
    print(f'eager float64 output: {eager_output}')

    torch.manual_seed(42)
    model_fp32 = copy.deepcopy(model)
    model_fp32 = model_fp32.float()
    model_fp32.eval()

    inputs_fp32 = tuple(t.float() if isinstance(t, torch.Tensor) and t.is_floating_point() else t for t in inputs)
    compiled_model = torch.compile(model_fp32, mode="reduce-overhead")

    with torch.no_grad():
        compiled_output = compiled_model(*inputs_fp32)
    print(f'compile float32 output: {compiled_output}')

    abs_diff = (eager_output - compiled_output.double()).abs()
    rel_diff = abs_diff / (eager_output.abs() + 1e-8)
    is_inconsistent = (abs_diff > 0.01) | (rel_diff > 0.001)

    print(f'Abs diff max: {abs_diff.max().item():.6e}, mean: {abs_diff.mean().item():.6e}')
    print(f'Rel diff max: {rel_diff.max().item():.6e}, mean: {rel_diff.mean().item():.6e}')
    print(f'Inconsistent: {is_inconsistent.sum().item()} / {eager_output.numel()}')

if __name__ == '__main__':
    main()
```

### Error logs

output:
```
eager float64 output: tensor([[[-1.3728,  0.3824,  1.8348,  ...,  0.6378, -0.1506,  1.1373],
         [ 0.0138,  0.0782, -1.4292,  ...,  0.6261, -1.5088, -1.1868],
         [ 2.3778,  1.4175, -1.1416,  ...,  1.5157, -0.3502, -1.8752],
         ...,
         [-0.2354,  0.9643, -0.8156,  ...,  0.6797,  2.4486, -1.9463],
         [-1.0628,  0.8263, -0.5272,  ...,  2.3872, -0.7048,  2.0775],
         [-1.1011,  0.0225, -0.4528,  ..., -0.8168,  0.8334, -0.9431]],

        [[ 0.9156,  0.0245,  0.1336,  ...,  1.0562,  0.0580,  2.9075],
         [-2.9002,  0.2870,  1.6100,  ...,  1.4399,  0.2331,  0.9262],
         [-0.7448,  1.0318,  0.5728,  ...,  0.8070, -0.3990,  0.4022],
         ...,
         [-0.6562, -1.0018,  0.3400,  ...,  0.1985,  0.0655,  1.5819],
         [-1.2791, -0.0341, -0.3818,  ..., -3.1152, -0.7449, -0.2718],
         [ 0.1153,  0.1189,  1.8075,  ..., -0.5691, -2.2755,  1.0768]]],
       dtype=torch.float64)
compile float32 output: tensor([[[-1.4080,  0.3511,  1.7842,  ...,  0.6192, -0.1069,  1.1530],
         [ 0.1526,  0.1074, -1.3940,  ...,  0.5715, -1.5751, -1.2773],
         [ 2.3747,  1.4169, -1.1494,  ...,  1.5233, -0.3552, -1.8707],
         ...,
         [-0.2224,  0.7930, -0.8713,  ...,  0.6338,  2.4210, -1.8844],
         [-1.1064,  1.0239, -0.4908,  ...,  2.3292, -0.4390,  1.9635],
         [-1.1379,  0.0433, -0.4131,  ..., -0.8359,  0.6859, -1.1654]],

        [[ 0.9159,  0.0248,  0.1332,  ...,  1.0569,  0.0579,  2.9075],
         [-2.9016,  0.2869,  1.6094,  ...,  1.4392,  0.2340,  0.9251],
         [-0.7441,  1.0318,  0.5725,  ...,  0.8068, -0.3994,  0.4029],
         ...,
         [-0.6561, -1.0017,  0.3402,  ...,  0.1983,  0.0655,  1.5819],
         [-1.2789, -0.0341, -0.3815,  ..., -3.1151, -0.7453, -0.2717],
         [ 0.1154,  0.1191,  1.8078,  ..., -0.5689, -2.2757,  1.0774]]])
Abs diff max: 1.288338e+00, mean: 3.672780e-02
Rel diff max: 1.758914e+03, mean: 2.629224e-01
Inconsistent: 11633 / 20480
```

### Versions

PyTorch version: 2.10.0+cu126

OS: Ubuntu 24.04.1 LTS

GPU: NVIDIA GeForce RTX 4060 Laptop GPU

CUDA version: 12.6

Python version: 3.12.3

cc @chauhang @penguinwu

**Comment by malfet** [comment]
Closing as won't-fix.

The repro compares an **eager float64** run against a **compiled float32** run (`model.double()` vs `model_fp32.float()`) and flags any deviation above a hand-picked `abs > 0.01 | rel > 0.001`. That threshold is tighter than float32 can deliver through a multi-layer attention chain, and the report gives no justification for why the float32 result is *wrong* rather than simply lower-precision. As the bug template asks: "If the difference is not expected as described in these documents, please provide appropriate justification on why one result is wrong and the other is correct." See [Numerical accuracy](https://docs.pytorch.org/docs/stable/notes/numerical_accuracy.html): "PyTorch is not guaranteed to produce bitwise identical results for floating point computations that are mathematically identical."

The reported magnitude is an amplification artifact rather than evidence of a compiler bug. The attention scores are computed as `attn_scores = self.head_dim ** 0.5 / (q @ k.transpose(-2, -1))` — a division by a matmul result that passes close to zero. Near zero, `1/x` turns an ordinary float32 rounding difference in `x` into an arbitrarily large difference, which then propagates through `softmax` and the output projection.

(Note that `nn.Dropout` in this model is inactive: `model.eval()` disables it on both paths, so the divergence here is dtype, not RNG.)

To show a genuine `torch.compile` correctness issue here, run both paths in the same dtype (float32 eager vs float32 compiled, same weights and inputs) and compare at default `rtol`/`atol`. #188425 was closed for the same reason.
